File: winnow/multicase.py
"""Reading OTHER cases, read-only, while one case stays open for writing.

An intrusion spans hosts, and a Winnow case is one host's evidence. The
questions an analyst actually has — "where else did this IOC land", "show
me these three machines on one timeline", "query two collections at once"
— all need several case files at once, and all of them are READS.

The design that follows from that: **one writable case, N read-only
readers.** The open case keeps its exclusive lock, its writer connection,
its views database and every invariant in CLAUDE.md exactly as written;
the others are opened `mode=ro`, with no lock, no views database and no
heartbeat, and nothing here can write to any of them.

That is not a limitation to work around later. Tags, notes and sessions
live in the case file that owns the rows, so writing across cases means
opening them read-write and fighting the one-Winnow-per-case lock. The
supported move is "jump to that row in its own case", which is a click.

Verified rather than assumed: a case that another Winnow has open and is
actively writing (WAL) reads fine through `mode=ro`, and SQLite attaches
at most 10 databases at once (`MAX_ATTACHED=10`) — which is why the
cross-case query path attaches in batches and everything else uses one
connection per case.
"""

from __future__ import annotations

import contextlib
import json
import os
import re
import sqlite3
from typing import Any, Iterator

# `v` is already attached to the writable Store's connection; leave room.
MAX_ATTACHED = 10
ATTACH_BUDGET = 8

# A table name out of a case's own `sources` row. Anchored because it is
# interpolated into SQL — the rows come from a file on disk, and a case
# file is data, not code (invariant #5's reasoning, applied across cases).
_TABLE_RE = re.compile(r"^src_\d+$")


class NotACaseFile(ValueError):
    """The path is readable but isn't a Winnow case."""


class CaseReader:
    """One other case, read-only. Close it, or use it as a context manager.

    Deliberately not a `Store`: Store takes the case lock, creates a views
    database and starts a heartbeat, all of which are exactly what a
    read-only visitor must not do. It is also a much smaller surface —
    everything here is a SELECT.
    """

    def __init__(self, path: str, name: str | None = None):
        self.path = str(path)
        self.name = name or os.path.splitext(os.path.basename(self.path))[0]
        if not os.path.isfile(self.path):
            raise NotACaseFile(f"No case file at {self.path}")
        self.db = sqlite3.connect(f"file:{self.path}?mode=ro", uri=True, check_same_thread=False)
        self.db.row_factory = sqlite3.Row
        try:
            names = {r[0] for r in self.db.execute(
                "SELECT name FROM sqlite_master WHERE type='table'")}
        except sqlite3.DatabaseError as e:      # not a database at all
            self.close()
            raise NotACaseFile(f"{self.path} is not readable as a case: {e}")
        if not {"sources", "row_tags"} <= names:
            self.close()
            raise NotACaseFile(f"{self.path} has no Winnow case tables")

    # ------------------------------------------------------------ lifecycle

    def close(self) -> None:
        with contextlib.suppress(Exception):
            self.db.close()

    def __enter__(self) -> "CaseReader":
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    # --------------------------------------------------------------- reads

    def sources(self) -> list[dict]:
        """Every real table in this case, with its columns parsed."""
        out = []
        for r in self.db.execute(
                "SELECT id, name, table_name, row_count, columns, nickname "
                "FROM sources ORDER BY id"):
            if not _TABLE_RE.match(r["table_name"] or ""):
                continue          # a case file naming something else is not ours to trust
            try:
                cols = json.loads(r["columns"] or "[]")
            except json.JSONDecodeError:
                cols = []
            out.append({"id": r["id"], "name": r["name"], "table_name": r["table_name"],
                        "row_count": r["row_count"], "nickname": r["nickname"],
                        "columns": [c["name"] for c in cols],
                        "column_types": {c["name"]: c.get("type") for c in cols}})
        return out

    def datetime_columns(self, source: dict) -> list[str]:
        return [c for c, t in source["column_types"].items() if t == "datetime"]

    def schema_text(self, alias: str) -> str:
        """CREATE TABLE-ish text for the SQL pane, qualified by the alias
        the caller attached this case under."""
        lines = [f"-- {self.name} ({os.path.basename(self.path)}) attached as {alias}"]
        for s in self.sources():
            cols = ",\n  ".join(f'"{c}" TEXT' for c in s["columns"])
            lines.append(f'CREATE TABLE {alias}."{s["table_name"]}" (  -- {s["name"]}'
                         f' · {s["row_count"]:,} rows\n  {cols}\n);')
        return "\n".join(lines)

    def watchlist_values(self) -> list[str]:
        if not self._has_table("watchlist"):
            return []
        return [r[0] for r in self.db.execute(
            "SELECT value FROM watchlist WHERE value <> '' ORDER BY value")]

    def _has_table(self, name: str) -> bool:
        return self.db.execute(
            "SELECT 1 FROM sqlite_master WHERE type='table' AND name=?", (name,)).fetchone() is not None

    # ------------------------------------------------------------ searching

    def find_values(self, values: list[str], *, limit_per_case: int = 200,
                    columns_per_row: int = 0) -> list[dict]:
        """Rows in this case containing any of `values`, as a substring, in
        any column. One hit per (source, rid, value) — the point is "where
        did this land", not a full row dump; `columns_per_row` > 0 asks for
        that many cells of context.

        LIKE, not the FTS index: a visiting reader must not depend on
        whether the other case ever built one, and must never build it.
        """
        wanted = [v for v in (values or []) if str(v).strip()]
        if not wanted:
            return []
        hits: list[dict] = []
        for src in self.sources():
            if len(hits) >= limit_per_case:
                break
            cols = src["columns"]
            if not cols:
                continue
            blob = " || '\\u0001' || ".join(f'COALESCE("{c}", \'\')' for c in cols)
            where = " OR ".join(f"{blob} LIKE ?" for _ in wanted)
            params = [f"%{v}%" for v in wanted]
            sel = "rid"
            if columns_per_row:
                sel += ", " + ", ".join(f'"{c}"' for c in cols[:columns_per_row])
            try:
                rows = self.db.execute(
                    f'SELECT {sel} FROM "{src["table_name"]}" WHERE {where} '
                    f"LIMIT {int(limit_per_case - len(hits))}", params).fetchall()
            except sqlite3.Error:
                continue          # a malformed source must not sink the sweep
            for r in rows:
                matched = [v for v in wanted
                           if any(v.lower() in str(x).lower() for x in tuple(r)[1:] or ())] if columns_per_row else []
                hits.append({
                    "case": self.name, "path": self.path,
                    "source_id": src["id"], "source": src["name"], "rid": r["rid"],
                    "values": matched or None,
                    "cells": {c: r[c] for c in cols[:columns_per_row]} if columns_per_row else {},
                })
        return hits

    # ------------------------------------------------------------- timeline

    def timeline_rows(self, *, ts_column: dict[int, str] | None = None,
                      start: str = "", end: str = "", limit: int = 5000) -> list[dict]:
        """Rows from this case as timeline entries: (when, case, source,
        rid, summary). `ts_column` picks the timestamp column per source id;
        without one, the first datetime column is used and a source with
        none is skipped."""
        out: list[dict] = []
        for src in self.sources():
            if len(out) >= limit:
                break
            col = (ts_column or {}).get(src["id"]) or next(iter(self.datetime_columns(src)), None)
            if not col or col not in src["columns"]:
                continue
            others = [c for c in src["columns"] if c != col][:4]
            sel = ", ".join(f'"{c}"' for c in [col] + others)
            where, params = [f'"{col}" <> \'\''], []
            if start:
                where.append(f'"{col}" >= ?'); params.append(start)
            if end:
                where.append(f'"{col}" <= ?'); params.append(end)
            try:
                rows = self.db.execute(
                    f'SELECT rid, {sel} FROM "{src["table_name"]}" WHERE {" AND ".join(where)} '
                    f'ORDER BY "{col}" LIMIT {int(limit - len(out))}', params).fetchall()
            except sqlite3.Error:
                continue
            for r in rows:
                out.append({
                    "when": r[col], "case": self.name, "path": self.path,
                    "source_id": src["id"], "source": src["name"], "rid": r["rid"],
                    "column": col,
                    "summary": " · ".join(str(r[c]) for c in others if r[c] not in (None, "")),
                })
        return out


@contextlib.contextmanager
def open_cases(paths: list[str], names: dict[str, str] | None = None) -> Iterator[list[CaseReader]]:
    """Open several cases read-only, closing every one of them even if a
    later open fails."""
    readers: list[CaseReader] = []
    try:
        for p in paths:
            readers.append(CaseReader(p, (names or {}).get(p)))
        yield readers
    finally:
        for r in readers:
            r.close()


# ------------------------------------------------------------- cross-case SQL

_FORBIDDEN_RE = re.compile(r"\b(attach|detach|pragma|vacuum)\b", re.IGNORECASE)


def _blank_string_literals(sql: str) -> str:
    """Single-quoted literals blanked, so a keyword inside a string is not
    read as a keyword. Same idea as Store's SQL-pane guard."""
    out, in_str = [], False
    for ch in sql:
        if ch == "'":
            in_str = not in_str
            out.append(" ")
        else:
            out.append(" " if in_str else ch)
    return "".join(out)


def query_across(paths: list[str], sql: str, *, limit: int = 5000,
                 names: dict[str, str] | None = None) -> dict:
    """Run ONE read-only SELECT across several cases, each attached under
    `c1`, `c2`, … in the order given.

    A fresh in-memory connection with the cases attached `mode=ro`, so the
    query cannot reach the open case's writer connection, cannot write
    anywhere, and disappears with the request. ATTACH/DETACH/PRAGMA/VACUUM
    are refused for the same reason Store.run_sql refuses them: the caller
    picks which cases are visible, not the SQL.
    """
    if not paths:
        raise ValueError("Pick at least one case")
    if len(paths) > ATTACH_BUDGET:
        raise ValueError(
            f"SQLite attaches at most {MAX_ATTACHED} databases at once — "
            f"pick {ATTACH_BUDGET} cases or fewer")
    structural = _blank_string_literals(sql or "")
    if not structural.strip():
        raise ValueError("Write a query")
    if _FORBIDDEN_RE.search(structural):
        raise ValueError("ATTACH, DETACH, PRAGMA and VACUUM aren't allowed here")

    with open_cases(paths, names) as readers:      # validates each is a case first
        aliases = [f"c{i + 1}" for i in range(len(readers))]
        conn = sqlite3.connect(":memory:")
        try:
            conn.row_factory = sqlite3.Row
            for alias, r in zip(aliases, readers):
                conn.execute(f"ATTACH DATABASE ? AS {alias}", (f"file:{r.path}?mode=ro",))
            import time as _time
            t0 = _time.time()
            cur = conn.execute(sql)
            rows = cur.fetchmany(limit)
            cols = [d[0] for d in cur.description] if cur.description else []
            return {
                "columns": cols,
                "rows": [[r[i] for i in range(len(cols))] for r in rows],
                "truncated": len(rows) == limit,
                "elapsed_ms": int((_time.time() - t0) * 1000),
                "cases": [{"alias": a, "name": r.name, "path": r.path}
                          for a, r in zip(aliases, readers)],
            }
        finally:
            conn.close()


def schema_across(paths: list[str], names: dict[str, str] | None = None) -> str:
    """The attached cases' schema, aliased the way query_across attaches
    them — what the SQL pane shows and what an LLM would be handed."""
    with open_cases(paths, names) as readers:
        return "\n\n".join(r.schema_text(f"c{i + 1}") for i, r in enumerate(readers))


def sweep_values(paths: list[str], values: list[str], *, limit_per_case: int = 200,
                 columns_per_row: int = 4, names: dict[str, str] | None = None) -> list[dict]:
    """IOC sweep: which of these cases hold any of these values, and where.

    One reader at a time rather than attaching, so the number of cases is
    not capped at SQLite's attach limit — an analyst with thirty hosts is
    the case this exists for. A case that fails to open is reported, not
    fatal: a sweep that stops at the first unreadable file is useless on a
    share where one case is mid-copy.
    """
    out = []
    for p in paths:
        try:
            with CaseReader(p, (names or {}).get(p)) as r:
                hits = r.find_values(values, limit_per_case=limit_per_case,
                                     columns_per_row=columns_per_row)
                out.append({"case": r.name, "path": p, "hits": hits,
                            "truncated": len(hits) >= limit_per_case, "error": None})
        except (NotACaseFile, sqlite3.Error) as e:
            out.append({"case": os.path.splitext(os.path.basename(p))[0], "path": p,
                        "hits": [], "truncated": False, "error": str(e)})
    return out


def timeline_across(paths: list[str], *, start: str = "", end: str = "",
                    limit: int = 5000, ts_columns: dict[str, dict[int, str]] | None = None,
                    names: dict[str, str] | None = None) -> dict:
    """One timeline over several cases, sorted by the timestamp text.

    Sorted in Python over each case's already-ordered rows rather than in
    SQL: the values are the canonical "YYYY-MM-DD HH:MM:SS[.ffffff]" shape
    (timeparse's whole point), so lexicographic order IS chronological, and
    this way the case count is not capped by SQLite's attach limit.

    Read-only by construction, and that is the design: tags and notes live
    in the case that owns the row, so the way to act on a hit is to open it
    in its own case.
    """
    rows: list[dict] = []
    errors = []
    for p in paths:
        try:
            with CaseReader(p, (names or {}).get(p)) as r:
                rows.extend(r.timeline_rows(
                    ts_column=(ts_columns or {}).get(p), start=start, end=end, limit=limit))
        except (NotACaseFile, sqlite3.Error) as e:
            errors.append({"path": p, "error": str(e)})
    rows.sort(key=lambda x: (x["when"] or "", x["case"], x["source_id"], x["rid"]))
    return {"rows": rows[:limit], "truncated": len(rows) > limit,
            "total_seen": len(rows), "errors": errors}

File: static/js/multicase.js
/* Across cases — one modal over the read-only cross-case routes.

   An intrusion spans hosts and a case is one host, so the questions that
   matter most are the ones a single case cannot answer: where else did
   this IOC land, show me these machines on one timeline, query two
   collections at once. Everything here is READ-ONLY by construction
   (winnow/multicase.py): the open case keeps its lock and its writer, the
   others are opened mode=ro. Acting on a hit means opening it in the case
   that owns it, which is what every row here offers.

   WITHDRAWN FROM THE HEADER, September 2026. The feature isn't finished
   enough to sit beside Search all: the three panes answer three
   different questions with three different result shapes, and the
   sweep's output stops short of the thing an analyst does next (tag the
   hits, pivot to the case). The routes, this module and its UI test all
   stay — openMultiCase() is still reachable via __winnow for whoever
   picks the redesign up — but nothing in index.html points here until
   it earns the space back. */
import { $, api, el, post, toast } from './core.js';
import { openCase } from './home.js';
import { markModalAction, modal } from './ui.js';

let picked = new Set();        // case paths, remembered while the modal is open
let mode = 'sweep';

function caseRow(c) {
  const row = el('label', 'mc-case');
  const cb = el('input');
  cb.type = 'checkbox';
  cb.checked = picked.has(c.path);
  cb.disabled = !c.exists;
  cb.onchange = () => { cb.checked ? picked.add(c.path) : picked.delete(c.path); };
  const name = el('span', 'mc-case-name', c.name);
  const note = el('span', 'mc-case-note',
    !c.exists ? 'file missing' : (c.is_open ? 'open here — read from this session' : (c.group || '')));
  row.append(cb, name, note);
  return row;
}

/* A hit or a timeline row: the case, where in it, and the way back. */
function resultRow(r, { time = false } = {}) {
  const row = el('div', 'mc-hit' + (time ? ' mc-hit-time' : ''));
  if (time) row.append(el('span', 'mc-when', r.when || ''));
  row.append(el('span', 'mc-case-tag', r.case));
  row.append(el('span', 'mc-src', `${r.source} · rid ${r.rid}`));
  const summary = time ? r.summary : Object.entries(r.cells || {})
    .map(([k, v]) => `${k}=${v}`).join('  ');
  row.append(el('span', 'mc-summary', summary));
  const go = el('button', 'btn ghost mc-open', 'Open');
  go.title = `Open ${r.case} at this row — tags and notes live in the case that owns it`;
  go.onclick = async () => {
    try {
      $('modal').hidden = true;
      await openCase(r.path);
      await api(`/api/source/${r.source_id}/open`, { method: 'POST' }).catch(() => {});
      toast(`Opened ${r.case} — ${r.source}, row ${r.rid}`, 5000);
    } catch (e) { toast('Could not open that case: ' + e.message, 6000); }
  };
  row.append(go);
  return row;
}

export function openMultiCase() {
  markModalAction('openMultiCase');
  modal('Across cases', async (b) => {
    b.append(el('p', 'fb-help',
      'Reads your other cases without opening them — they stay untouched, and a case another '
      + 'Winnow has open is still readable. Everything here is read-only: to tag or annotate a '
      + 'row, open it in the case that owns it.'));

    const caseBox = el('div', 'mc-cases');
    b.append(el('div', 'settings-sub-label', 'Cases'), caseBox);

    let budget = 8;
    try {
      const info = await api('/api/multicase/cases');
      budget = info.attach_budget;
      if (!info.cases.length) caseBox.append(el('div', 'note-status', 'No cases registered yet.'));
      // Default to everything that exists, which is the sweep's best default.
      if (!picked.size) for (const c of info.cases) if (c.exists) picked.add(c.path);
      for (const c of info.cases) caseBox.append(caseRow(c));
    } catch (e) {
      caseBox.append(el('div', 'note-status', 'Could not read the case list: ' + e.message));
    }

    const tabs = el('div', 'row-actions mc-tabs');
    const out = el('div', 'mc-out');
    const controls = el('div', 'mc-controls');
    b.append(tabs, controls, out);

    const status = (msg) => out.replaceChildren(el('div', 'note-status', msg));

    // ---------------------------------------------------------- A: sweep
    async function runSweep(values) {
      status('Looking…');
      try {
        const r = await post('/api/multicase/sweep',
          { paths: [...picked], values: values || null });
        out.replaceChildren();
        out.append(el('div', 'fb-help',
          `${r.total_hits} hit${r.total_hits === 1 ? '' : 's'} for ${r.values.length} value`
          + `${r.values.length === 1 ? '' : 's'} across ${r.cases.length} case`
          + `${r.cases.length === 1 ? '' : 's'}`));
        for (const c of r.cases) {
          if (c.error) { out.append(el('div', 'mc-case-err', `${c.case}: ${c.error}`)); continue; }
          if (!c.hits.length) continue;
          out.append(el('div', 'mc-case-head',
            `${c.case} — ${c.hits.length}${c.truncated ? '+' : ''}`));
          for (const h of c.hits) out.append(resultRow(h));
        }
        if (!r.total_hits) out.append(el('div', 'note-status', 'Nothing found in the other cases.'));
      } catch (e) { status('Could not sweep: ' + e.message); }
    }

    // ------------------------------------------------------------- B: sql
    async function runSql(sql) {
      status('Running…');
      try {
        const r = await post('/api/multicase/sql', { paths: [...picked], sql });
        out.replaceChildren();
        out.append(el('div', 'fb-help',
          r.cases.map((c) => `${c.alias} = ${c.name}`).join(' · ')
          + ` — ${r.rows.length}${r.truncated ? '+' : ''} rows in ${r.elapsed_ms}ms`));
        const table = el('table', 'mc-table');
        const head = el('tr');
        for (const c of r.columns) head.append(el('th', null, c));
        table.append(head);
        for (const row of r.rows) {
          const tr = el('tr');
          for (const cell of row) tr.append(el('td', null, cell == null ? '' : String(cell)));
          table.append(tr);
        }
        out.append(table);
      } catch (e) { status(e.message); }
    }

    // -------------------------------------------------------- C: timeline
    async function runTimeline(start, end) {
      status('Building…');
      try {
        const r = await post('/api/multicase/timeline', { paths: [...picked], start, end });
        out.replaceChildren();
        out.append(el('div', 'fb-help',
          `${r.rows.length}${r.truncated ? ` of ${r.total_seen}+` : ''} rows`
          + (r.errors.length ? ` · ${r.errors.length} case(s) unreadable` : '')));
        for (const row of r.rows) out.append(resultRow(row, { time: true }));
        if (!r.rows.length) out.append(el('div', 'note-status',
          'Nothing in that window — these cases may have no datetime columns.'));
      } catch (e) { status('Could not build it: ' + e.message); }
    }

    function paint() {
      controls.replaceChildren();
      out.replaceChildren();
      if (mode === 'sweep') {
        controls.append(el('p', 'fb-help',
          'Blank uses this case’s watchlist — the IOCs you have already written down.'));
        const box = el('textarea', 'mc-values');
        box.rows = 2;
        box.placeholder = 'One value per line (blank = this case’s watchlist)';
        const go = el('button', 'btn', 'Sweep');
        go.onclick = () => runSweep(box.value.split('\n').map((v) => v.trim()).filter(Boolean));
        controls.append(box, go);
        controls.dataset.mode = 'sweep';
      } else if (mode === 'sql') {
        const box = el('textarea', 'mc-sql');
        box.rows = 4;
        box.placeholder = 'SELECT * FROM c1.src_1 LIMIT 50';
        const go = el('button', 'btn', 'Run');
        go.onclick = () => runSql(box.value);
        const schema = el('button', 'btn ghost', 'Show schema');
        schema.title = 'Cases attach as c1, c2… — they are schema names, so tables still need their own alias';
        schema.onclick = async () => {
          try {
            const r = await post('/api/multicase/schema', { paths: [...picked] });
            out.replaceChildren(el('pre', 'mc-schema', r.schema));
          } catch (e) { status(e.message); }
        };
        const acts = el('div', 'row-actions');
        acts.append(go, schema);          // el(tag, cls, TEXT) — children get appended
        const note = el('p', 'fb-help',
          `Up to ${budget} cases at once, attached as c1, c2… in the order listed above.`);
        controls.append(note, box, acts);
        controls.dataset.mode = 'sql';
        // SQLite attaches at most 10 databases, so this is the one panel
        // with a case limit. Say so before the query fails, and say which
        // ones would be used rather than making them count checkboxes.
        if (picked.size > budget) {
          note.classList.add('mc-over-budget');
          note.textContent = `${picked.size} cases selected — a query can attach at most ${budget}. `
            + 'Deselect some above, or the run will be refused.';
        }
      } else {
        const from = el('input', 'confirm-input');
        from.placeholder = 'from  YYYY-MM-DD HH:MM:SS';
        const to = el('input', 'confirm-input');
        to.placeholder = 'to';
        const go = el('button', 'btn', 'Build');
        go.onclick = () => runTimeline(from.value.trim(), to.value.trim());
        const acts = el('div', 'row-actions');
        acts.append(from, to, go);
        controls.append(el('p', 'fb-help',
          'Every case’s first datetime column, interleaved in time order.'), acts);
        controls.dataset.mode = 'timeline';
      }
    }

    // Switching tabs repaints the controls in place. Re-opening the modal
    // would re-fetch the case list and lose the analyst's selection, for a
    // change that only affects the panel below it.
    const tabButtons = [];
    for (const [id, label] of [['sweep', 'IOC sweep'], ['sql', 'Query'], ['timeline', 'Timeline']]) {
      const t = el('button', 'btn', label);
      t.dataset.mode = id;
      t.onclick = () => {
        mode = id;
        for (const b2 of tabButtons) b2.className = 'btn' + (b2.dataset.mode === mode ? '' : ' ghost');
        paint();
      };
      tabButtons.push(t);
      tabs.append(t);
    }
    for (const t of tabButtons) t.className = 'btn' + (t.dataset.mode === mode ? '' : ' ghost');
    paint();
  }, { wide: 'x' });
}
